src/gss_names.c

```diff
@@ -626,9 +708,104 @@ uint32_t gssntlm_inquire_name(uint32_t *minor_status,
                               gss_OID *MN_mech,
                               gss_buffer_set_t *attrs)
 {
-    return GSS_S_UNAVAILABLE;
+    uint32_t retmin;
+    uint32_t retmaj;
+    uint32_t tmpmin;
+    gss_buffer_desc buf;
+    const struct gssntlm_name *in = (const struct gssntlm_name *)name;
+
+    if (name == GSS_C_NO_NAME) {
+        return GSSERRS(GSS_S_BAD_NAME, GSS_S_CALL_INACCESSIBLE_READ);
+    }
+    if (!attrs) {
+        return GSSERRS(ERR_NOARG, GSS_S_CALL_INACCESSIBLE_WRITE);
+    }
+    *attrs = GSS_C_NO_BUFFER_SET;
+
+    if (in->attrs == NULL) {
+        return GSSERRS(0, GSS_S_COMPLETE);
+    }
+
+    for (size_t i = 0; in->attrs[i].attr_name != NULL; i++) {
+        struct gssntlm_name_attribute *attr = &in->attrs[i];
+        size_t attr_name_len = strlen(attr->attr_name);
+
+        gss_buffer_t attr_value = &attr->attr_value;
+        /* +1 for '=' separator and +1 for EOL */
+        buf.length = attr_value->length + attr_name_len + 2;
+        buf.value = malloc(buf.length);
+        if (buf.value == NULL) {
+            retmin = ENOMEM;
+            goto done;
+        }
+
+        /* Construct 'attr_name=<attr_value>\0' string */
+        memcpy(buf.value, attr->attr_name, attr_name_len);
+        ((char*)buf.value)[attr_name_len] = '=';
+        memcpy(((char*)buf.value) + attr_name_len + 1, attr_value->value, attr_value->length);
+        ((char*)buf.value)[buf.length] = 0;
+
+        retmaj = gss_add_buffer_set_member(&retmin, &buf, attrs);
+        free(buf.value);
+        if (retmaj != GSS_S_COMPLETE) goto done;
+    }
+    return GSSERRS(0, GSS_S_COMPLETE);
+
+done:
+    (void)gss_release_buffer_set(&tmpmin, attrs);
+    return GSSERRS(retmin, retmaj);
```

Comment:
`retmaj` may not be initialized here - e.g., if you came from line 739.  `retmin` may also not be - e.g., if you came from line 750.

Reply:
As for line 739 - yes, I see you use set_GSSERR() macro in such case. Using the macros makes the code more hard to read (because you don't see what you do), but since this is your approach - I will stick to it.

As for line 750 - we just propagate values of retmaj/retmin from gss_add_buffer_set_member(). If for some reason retmin is not set inside - this is not our fault. Do you think we have to do something else?

```
        retmaj = gss_add_buffer_set_member(&retmin, &buf, attrs);
        if (retmaj != GSS_S_COMPLETE) goto done;
```

Reply:
The problem is not in the gss_add_buffer_set_member() in thi case, the problem is that when you check for buf.value allocation (line 737) you set only retmin and then you goto out. When you do that retmaj has not yet been set, so it will have an arbitrary value, potentially 0.
the reason why I have the GSSERRS macro is exactly to not forget to set both values when I want a quick out. The macro is not necessary when a gss_something function is called because they always set both retmaj and retmin.

Reply:
I fixed allocation code to use set_GSSERR() and switched to 'return GSSERR()' in previous commit - is it good enough?